### PR TITLE
auto_lint

### PR DESCRIPTION
**Pull request by krober10nd**
#86

### CODE REVIEW

**Comment by krober10nd** [comment]
@zacharyburnettNOAA I'm trying to address #91 but by having the github-action bot automatically format the code for you. Needs some more work I guess.

**Comment by ghost** [comment]
@krober10nd this seems like a good idea on paper, though will it submit its own commits under its own sign? I'm wary that might clutter the commit history if every single push had a corresponding "formatting" commit alongside it

**Comment by krober10nd** [comment]
@zacharyburnettNOAA  Normally you squash commits in a PR before merging with main

**Comment by jreniel** [comment]
> @krober10nd this seems like a good idea on paper, though will it submit its own commits under its own sign? I'm wary that might clutter the commit history if every single push had a corresponding "formatting" commit alongside it

This is the reason why so far I chose to ignore linter differences, because I don't think exact uniform formatting is too important. But if we insist on having uniform formatting, then definitely a GitHub action is the price to pay. It won't work to try to convince other developers to install and use a specific linter, in particular because a lot of people simply "don't like" the way x or y linter "fixes" their code. By having an automated action, we simply get rid of people opinion in favor to the computers opinion.

**Comment by ghost** [comment]
> @zacharyburnettNOAA Normally you squash commits in a PR before merging with main

That makes sense, so we should make a new workflow YML just for pull requests that applies the formatting upon pull request creation / approval, maybe call it `.github/workflows/format.yml`.

https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request

Also, the main reason I like to use OITNB over Black is because it uses single quotes `'` instead of double quotes `"` and I prefer that. However, this isn't a huge deal if the github action only uses Black.

**Comment by jreniel** [comment]
> > @zacharyburnettNOAA Normally you squash commits in a PR before merging with main
> 
> That makes sense, so we should make a new workflow YML just for pull requests that applies the formatting, maybe call it `.github/workflows/format.yml`.
> 
> Also, the main reason I like to use OITNB over Black is because it uses single quotes `'` instead of double quotes `"` and I prefer that. However, this isn't a huge deal if the github action only uses Black.

That's the price to pay if we want to have uniform formatting. Not everyone will be happy with the result, but everyone needs to accept whatever the computer spits out. This is why I didn't bother with code formatting rules before, they are too subjective...

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+name: formatting
+
+on: pull_request
```
Comment (ghost) [inline .github/workflows/lint.yml]: only activate when pull request is created or committed to

File: .github/workflows/lint.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      - name: Setup Python
+        uses: actions/setup-python@v1
+      - name: Install dependencies
+        run: pip install black flake8
+      - name: Run linters / formatters
+        uses: wearerequired/lint-action@v1
+        with:
+          black: true
```
Comment (ghost) [inline .github/workflows/lint.yml]: moved this from `tests.yml`

Comment (ghost) [reply .github/workflows/lint.yml]: also removed `flake8` as it is already checked in `tests.yml`

File: .github/workflows/lint.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+          auto_fix: true
```
Comment (ghost) [inline .github/workflows/lint.yml]: this setting commits the `black` changes automatically

**Comment by krober10nd** [comment]
Great work @zacharyburnettNOAA

**Comment by jreniel** [comment]
> Great work @zacharyburnettNOAA

Agreed!

**Comment by ghost** [comment]
@krober10nd @JaimeCalzadaNOAA no problem, I'll go ahead and merge this in when the checks finish.

**Comment by krober10nd** [comment]
Tests won’t pass until #92 is resolved

**Comment by ghost** [comment]
> Tests won’t pass until #92 is resolved

here's the test, for reference:
```
======================================================================
ERROR: test_best_track_run (tests.test_best_track_run.BestTrackRunCliTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/adcircpy/adcircpy/tests/test_best_track_run.py", line 50, in test_best_track_run
    best_track_run.main()
  File "/home/runner/work/adcircpy/adcircpy/adcircpy/cmd/best_track_run.py", line 56, in main
    BestTrackRunCommand(args).run()
  File "/home/runner/work/adcircpy/adcircpy/adcircpy/cmd/best_track_run.py", line 24, in __init__
    bt.clip_to_bbox(self.mesh.get_bbox(output_type="bbox"), self.mesh.crs)
  File "/home/runner/work/adcircpy/adcircpy/adcircpy/forcing/winds/best_track.py", line 520, in clip_to_bbox
    ellps={"GRS 1980": "GRS80", "WGS 84": "WGS84"}[df_crs.ellipsoid.name],
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/pyproj/crs/crs.py", line 313, in __init__
    self._local.crs = _CRS(self.srs)
  File "pyproj/_crs.pyx", line 2326, in pyproj._crs._CRS.__init__
pyproj.exceptions.CRSError: Invalid projection: +proj=utm +zone=18P +ellps=WGS84 +type=crs: (Internal Proj Error: proj_create: Error 1027 (Invalid value for an argument): utm: Invalid value for zone)
```

additionally, this test seems to also be failing:
```
======================================================================
ERROR: test_external_bt (tests.test_besttrack.TestBestTrack)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/adcircpy/adcircpy/tests/test_besttrack.py", line 38, in test_external_bt
    bt.plot_track(show=True)
  File "/home/runner/work/adcircpy/adcircpy/adcircpy/forcing/winds/best_track.py", line 565, in plot_track
    _fetch_and_plot_coastline(axes, show)
  File "/home/runner/work/adcircpy/adcircpy/adcircpy/forcing/winds/best_track.py", line 36, in _fetch_and_plot_coastline
    save_dir.mkdir(exist_ok=True)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/pathlib.py", line 1273, in mkdir
    self._accessor.mkdir(self, mode)
FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/.local/share/ne_coastline'

----------------------------------------------------------------------
```

**Comment by krober10nd** [comment]
@zacharyburnettNOAA the second problem was fixed in #90. The other problem was temporarily fixed by pinning the version as discussed also in #90

**Comment by ghost** [comment]
`oitnb` support was added in https://github.com/wearerequired/lint-action/pull/242, so maybe we can look into using that. It really just depends on whether we want to use single quotes or double quotes in strings. I prefer single quotes, but it really doesn't matter too much as long as we enforce a standard

**Comment by jreniel** [comment]
> `oitnb` support was added in [wearerequired/lint-action#242](https://github.com/wearerequired/lint-action/pull/242), so maybe we can look into using that. It really just depends on whether we want to use single quotes or double quotes in strings. I prefer single quotes, but it really doesn't matter too much as long as we enforce a standard

I don't really care about any linter in particular as long as there are no justifiable objections by others.

**Comment by ghost** [comment]
alright, I'll switch to `oitnb` and add the badge to the README

**Pull request merged**